src/main/scala/com/crobox/clickhouse/stream/ClickhouseIndexingSubscriber.scala

```diff
@@ -279,35 +295,70 @@ class ClickhouseBulkActor(targetTable: String,
     buffer.clear()
     context.stop(self)
   }
-
-  val insertQuery = s"INSERT INTO $targetTable FORMAT JSONEachRow"
-
   private val objectMapper = new ObjectMapper()
     .registerModule(new JodaModule)
     .registerModule(new DefaultScalaModule)
     .setSerializationInclusion(JsonInclude.Include.NON_DEFAULT)
 
   private def index(): Unit = {
-    val payload = buffer.map(i => objectMapper.writeValueAsString(i.data)).mkString("\n") + "\n"
-    val count   = buffer.size
+    val colList = buffer
+      .map(_.data)
+      .reduce(
+        (e1, e2) => if (e1.size > e2.size) e1 else e2
+      )
+      .keys
+      .toSeq
+
+    val regularPayload = buffer
+      .collect {
+        case InsertParsed(table, data) => {
+          "(" + data.values.mkString(", ") + ")"
+        }
+      }
+      .mkString(",\n") + "\n"
+
+    val jsonPayload = buffer
+      .collect {
+        case InsertRaw(table, data) => objectMapper.writeValueAsString(data)
+      }
+      .mkString("\n") + "\n"
+
+    val count = buffer.size
     logger.debug(s"Inserting $count")
+
     sent = sent + count
     logger.debug(s"Clickhouse $targetTable sent: $sent (confirmed: $confirmed, failed: $failed)")
-    send(payload, count)
+    send(targetTable, jsonPayload, count)
+    send(targetTable, regularPayload, count, rawColList = colList)
     buffer.clear()
 
     // buffer is now empty so no point keeping a scheduled flush after operation
     flushAfterScheduler.foreach(_.cancel)
     flushAfterScheduler = None
   }
 
-  private def send(payload: String, count: Int, retries: Int = 3): Unit =
+  private def send(table: String,
+                   payload: String,
+                   count: Int,
+                   retries: Int = 3,
+                   rawColList: Seq[String] = Seq.empty): Unit = {
+    val insertQuery = {
+      if (rawColList.isEmpty) {
+        s"INSERT INTO $table FORMAT JSONEachRow"
+      } else {
+        val colList = rawColList.mkString(", ")
+        s"INSERT INTO $table ($colList) VALUES"
+      }
+    }
     client.execute(insertQuery, payload) onComplete {
-      case Failure(_) if retries > 0 => send(payload, count, retries - 1)
-      case Failure(e)                => self ! ClickhouseBulkActor.FlushFailure(e, count)
-      case Success(resp: String)     => self ! ClickhouseBulkActor.FlushSuccess(resp, count)
+      case Failure(_) if retries > 0 =>
+        send(table, payload, count, retries - 1, rawColList)
+      case Failure(e) =>
```

Comment:
Isn't the retry logic present in the executor as well? Could we reuse it somehow?
Also I do not think that all inserts are retriable, we could retry network failures or unexpected closed connections but should not retry all failures

Reply:
yeah, altho its not really in the scope of this PR... we can make a task for it internally.
Id rather spend more effort on refactoring the clickhouse client when we get to set up the DSL project separately, for then the SOC becomes more apparent.

Reply:
retry logic is present inside the clickhouse client so it is not necesarry here. I took it out